### ISSUE TITLE
About PPO training

### ISSUE THREAD

**Issue by Pefect96** [CuthbertCai/Ask-Confirm#2]
How is your PPO trained? Does the code seem to train only one sample for each epoch? How do you embed your AC model in S-SCAN? As far as I know, SCAN is a method of image-text matching using caption, while your AC uses 1600 Objects. I don't quite understand this, and I hope you can help me.

**Comment by CuthbertCai** [comment]
For each epoch, one image is sampled, and in several round interactions, objects are suggested by the policy net to enrich the initial caption. During the interaction, the action, state, policy, value and reward per round are stored to optimize policy and value nets with PPO. In another word, although only one image is sampled, the action, state, policy, value and reward change each round.

In our method, AC exactly should work with an image-text matching model, because the goal of AC is to improve image-text retrieval performance. AC tries to append some captions of objects to the initial caption so that the image-text matching model (e.g., SCAN) can better retrieve the target image with the reformed caption.  For example, if the initial caption is "a woman stands by the street", AC may suggest that "sky, tree, table" are possible to show in the image, then the reformed caption is "a woman stands by the street. sky, tree, table" and SCAN could get better retrieval results with such reformed caption.

**Comment by Pefect96** [comment]
> For each epoch, one image is sampled, and in several round interactions, objects are suggested by the policy net to enrich the initial caption. During the interaction, the action, state, policy, value and reward per round are stored to optimize policy and value nets with PPO. In another word, although only one image is sampled, the action, state, policy, value and reward change each round.
> 
> In our method, AC exactly should work with an image-text matching model, because the goal of AC is to improve image-text retrieval performance. AC tries to append some captions of objects to the initial caption so that the image-text matching model (e.g., SCAN) can better retrieve the target image with the reformed caption. For example, if the initial caption is "a woman stands by the street", AC may suggest that "sky, tree, table" are possible to show in the image, then the reformed caption is "a woman stands by the street. sky, tree, table" and SCAN could get better retrieval results with such reformed caption.

- Thank you for your reply. You mean AC is trained on the basis of trained image retrieval models, such as S-SCAN and T-CMPL. The basic model is already capable of retrieving images by caption. In addition, AC can make the image and text retrieval results of S-SCAN and T-CMPL models more accurate. In other words, according to Figure 4 in your paper and supplementary materials, AC can advance the ranking of target image retrieval. I don't know if my understanding is correct.
- In SCAN, the original author appears to use the MSCOCO dataset, why you use VG.  
- However, In your paper, the number of training data is 92105, does it only need 500 epoch to be convergent and effective? I feel it is not enough, because only 500 training samples are used to interactively train AC. According to my understanding of reinforcement learning, reinforcement learning algorithms are very difficult to converge. Your script (script/train_ppo_coherence.sh) seems to have 5000 epochs to train the AC, but even 5,000 , I feel it is not enough.
- I do not know whether my understanding is correct, I hope you can teach me, if I say wrong, please forgive me.

**Comment by CuthbertCai** [comment]
* For the first point, yes, you are right.
* The reasons why we use VG instead of MSCOCO include: 1) object categories of MSCOCO are too few. We want to expand our method to a more general setting, thus VG that consists of 1600 object categories is more suitable; 2) considering we focus on the interactive retrieval problem and the previous work, DrillDown, has already adopted VG, we follow DrillDown's setting.
* RL is indeed very difficult to converge. According to our experiments, if we remove the shaping method we proposed in AC, the PPO can not even converge, because the samples are really limited. Our shaping method pre-computes some statistics of the whole image gallery. The policy net is forced to fit the pre-computed statistics and performs PPO optimization simultaneously so that the training could converge. The reason why we only need to sample a subset of the dataset to perform PPO is that the pre-computed statistics already cover the whole dataset, the RL part only needs to fine-tune the pre-computed statistics. The RL part is also very important. If the model is only supervised by the pre-computed statistics, only the most frequent objects are selected such as "sky, tree, people", whereas RL could find some rarely appeared objects such as "trophy". But we also agree that training 500 epochs is a little bit counterintuitive, the reason for it is that training more epochs (> 500) cannot get better results according to our experiments. 

Thanks for your comment! Hope the information above could help you.

**Comment by Pefect96** [comment]
- Thank you for your help. I learned a lot through your explanation. So AC works really well and maybe that's because Shaping is so powerful. 
- May I ask how you train S-SCAN and T-CMPL? The title of the image or the title of the region? If the title of the image is used, why test with the title of the region? I am not sure of the significance of training these two image-text matching model  in advance.

**Comment by CuthbertCai** [comment]
The title of regions is used both in the training and testing periods. 

We train these image-text matching models in advance because we want to make sure the performance gain is achieved by the proposed interactive retrieval framework. If we also fine-tune the image-text matching model, the improvement may be caused by other reasons, such as more training epochs or other hyper-parameters. 

Another reason is that in practice, to solve a retrieval task, we usually tend to directly train an image-text matching model. Once we already have the well-trained model, to modify it into an interactive retrieval model,  by using our method, we can only train AC based on the well-trained model instead of training an interactive model from scratch. This setting may make our method generalize to more scenarios.

**Comment by Pefect96** [comment]
Thank you very much for your help. I'll take a closer look.
